google-cloud-tools-plugin/jps-plugin/src/com/google/cloud/tools/intellij/jps/GenRepoInfoFileModuleBuilder.java

```diff
@@ -0,0 +1,169 @@
+ */
+public class GenRepoInfoFileModuleBuilder extends ModuleLevelBuilder {
+
+  private static final Logger LOG =
+      Logger.getInstance("#com.google.cloud.tools.intellij.jps.GenRepoInfoFileModuleBuilder");
+
+  public static final String NAME = "Stackdriver source context generator";
+  private GenRepoInfoFileActionFactory actionFactory = new GenRepoInfoFileActionFactory();
+
+  public GenRepoInfoFileModuleBuilder() {
+    super(BuilderCategory.CLASS_POST_PROCESSOR);
+  }
+
+  @VisibleForTesting
+  GenRepoInfoFileModuleBuilder(GenRepoInfoFileActionFactory actionFactory) {
+    this();
+    this.actionFactory = actionFactory;
+  }
+
+  /**
+   * Generates source context files for each module in {@code chunk}.
+   *
+   * <p>Generation is parameterized by the Stackdriver facet associated to the module.
+   *
+   * <p>Files are added to the same location as the project's compiled .class files.
+   */
+  @Override
+  public ExitCode build(CompileContext context, ModuleChunk chunk,
+      DirtyFilesHolder<JavaSourceRootDescriptor, ModuleBuildTarget> dirtyFilesHolder,
+      OutputConsumer outputConsumer) throws ProjectBuildException, IOException {
+    for (JpsModule jpsModule : chunk.getModules()) {
+      JpsStackdriverModuleExtension extension = jpsModule.getContainer().getChild(
+          JpsStackdriverModuleExtensionImpl.ROLE);
+
+      if (!extension.isGenerateSourceContext()) {
+        continue;
+      }
+
+      if (jpsModule.getSourceRoots().isEmpty()) {
+        LOG.warn("Module " + jpsModule.getName() + " contains no source roots. Moving on to the "
+            + "next module for source generation.");
+        continue;
+      }
+
+      if (extension.getCloudSdkPath() == null) {
+        LOG.warn("No Cloud SDK path specified. Skipping source context generation.");
+        // Cloud SDK path is a singleton for a project, so if it doesn't exist on the first module,
+        // we can return right away.
+        return ExitCode.NOTHING_DONE;
+      }
+
+      if (extension.getModuleSourceDirectory() == null) {
+        // Possible to happen on Windows, unlikely in Linux.
+        LOG.warn("Invalid module source directory. Check for special characters <>:\"|?*.");
+        continue;
+      }
+
+      Path sourceDirectory = extension.getModuleSourceDirectory();
+      ModuleBuildTarget target =
+          new ModuleBuildTarget(jpsModule, JavaModuleBuildTargetType.PRODUCTION);
+      Path outputDirectory = target.getOutputDir().toPath();
+
+      GenRepoInfoFile genAction = actionFactory.newAction(extension.getCloudSdkPath());
+
+      DefaultGenRepoInfoFileConfiguration configuration = new DefaultGenRepoInfoFileConfiguration();
+      configuration.setSourceDirectory(sourceDirectory.toFile());
+      configuration.setOutputDirectory(outputDirectory.toFile());
+      genAction.generate(configuration);
+
+      ExitCodeRecorderProcessExitListener exitListener = actionFactory.getExitListener();
```

Comment:
We're calling generate() synchronously. ProcessExitListeners happen to be the only way to retrieve gcloud's exit code from the common lib.

Are you saying we should make `generate()` retrieve the status code? There might be other methods where we want to return something else, and we wouldn't be able to stick to this convention. In any case, I think any change here could be handled outside of this PR?